ext/uri/php_uri.c

```diff
@@ -40,12 +48,565 @@ static const zend_module_dep uri_deps[] = {
+	bool should_throw, bool should_update_this_object, zval *errors_zv
+) {
+	zval errors;
+	ZVAL_UNDEF(&errors);
+
+	void *base_url = NULL;
+	if (base_url_object != NULL) {
+		uri_internal_t *internal_base_url = uri_internal_from_obj(base_url_object);
+		URI_ASSERT_INITIALIZATION(internal_base_url);
+		base_url = internal_base_url->uri;
+	}
+
+	void *uri = handler->parse_uri(uri_str, base_url, should_throw || errors_zv != NULL ? &errors : NULL);
+	if (UNEXPECTED(uri == NULL)) {
+		if (should_throw) {
+			throw_invalid_uri_exception(handler, &errors);
+			zval_ptr_dtor(&errors);
+			RETURN_THROWS();
+		} else {
+			pass_errors_by_ref(errors_zv, &errors);
+			zval_ptr_dtor(&errors);
+			RETURN_NULL();
+		}
+	}
+
+	pass_errors_by_ref(errors_zv, &errors);
+	zval_ptr_dtor(&errors);
+
+	uri_object_t *uri_object;
+	if (should_update_this_object) {
+		uri_object = Z_URI_OBJECT_P(ZEND_THIS);
+	} else {
+		if (EX(func)->common.fn_flags & ZEND_ACC_STATIC) {
+			object_init_ex(return_value, Z_CE_P(ZEND_THIS));
+		} else {
+			object_init_ex(return_value, Z_OBJCE_P(ZEND_THIS));
+		}
+		uri_object = Z_URI_OBJECT_P(return_value);
+	}
+
+	uri_object->internal.handler = handler;
+	uri_object->internal.uri = uri;
+}
+
+static void create_whatwg_uri(INTERNAL_FUNCTION_PARAMETERS, bool is_constructor)
+{
+	zend_string *uri_str;
+	zend_object *base_url_object = NULL;
+	zval *errors = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_PATH_STR(uri_str)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_NULL(base_url_object, whatwg_url_ce)
+		Z_PARAM_ZVAL(errors)
+	ZEND_PARSE_PARAMETERS_END();
+
+	php_uri_instantiate_uri(INTERNAL_FUNCTION_PARAM_PASSTHRU, &lexbor_uri_handler, uri_str, base_url_object, is_constructor, is_constructor, errors);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, parse)
+{
+	create_whatwg_uri(INTERNAL_FUNCTION_PARAM_PASSTHRU, false);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, __construct)
+{
+	create_whatwg_uri(INTERNAL_FUNCTION_PARAM_PASSTHRU, true);
+}
+
+static void uri_equals(INTERNAL_FUNCTION_PARAMETERS, zend_object *that_object, zend_object *comparison_mode)
+{
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *this_internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(this_internal_uri);
+
+	uri_internal_t *that_internal_uri = uri_internal_from_obj(that_object);
+	URI_ASSERT_INITIALIZATION(that_internal_uri);
+
+	if (this_object->ce != that_object->ce &&
+		!instanceof_function(this_object->ce, that_object->ce) &&
+		!instanceof_function(that_object->ce, this_object->ce)
+	) {
+		RETURN_FALSE;
+	}
+
+	bool exclude_fragment = true;
+	if (comparison_mode) {
+		zval *case_name = zend_enum_fetch_case_name(comparison_mode);
+		zend_string *comparison_mode_name = Z_STR_P(case_name);
+
+		exclude_fragment = ZSTR_VAL(comparison_mode_name)[0] + ZSTR_LEN(comparison_mode_name) == 'E' + sizeof("ExcludeFragment") - 1;
```

Comment:
Since we are only comparing between two cases this can just be `zend_string_equals_literal(Z_STR_P(case_name), "ExcludeFragment")`.